File: app/src/main/java/sakaitakao/android/permissionsearch/activity/PermissionListActivity.java
package sakaitakao.android.permissionsearch.activity;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

import sakaitakao.android.permissionsearch.R;
import sakaitakao.android.permissionsearch.adaptor.PermissionListAdaptor;
import sakaitakao.android.permissionsearch.entity.PermissionInfoEx;
import sakaitakao.android.permissionsearch.model.AppPermissionsFinder;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;
import android.widget.ListView;
import android.widget.TextView;

/**
 * 権限一覧のアクティビティ
 * 
 * @author takao
 * 
 */
public class PermissionListActivity extends Activity {

	private static final String INTENTEXTRA_CONDITION = PermissionListActivity.class.getName() + ".CONDITION";

	private PermissionListAdaptor adaptor;

	/** Called when the activity is first created. */
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.permission_list);

		SearchCondition condition = getCondition();

		PackageManager packageManager = getPackageManager();
		AppPermissionsFinder aph = new AppPermissionsFinder();
		List<PermissionInfoEx> list = aph.getPermissionInfoExList(packageManager, toAppPermissionsFinderCondition(condition));

		Resources resources = getResources();
		showHitCount(resources, list);
		showList(list);
	}

	/**
	 * アクティビティ起動時の検索条件を設定する。
	 * 
	 * @param intent
	 *            {@link Intent}
	 * @param searchCondition
	 *            検索条件
	 */
	public static void setSearchCondition(Intent intent, SearchCondition searchCondition) {
		intent.putExtra(PermissionListActivity.INTENTEXTRA_CONDITION, searchCondition);
	}

	// -------------------------------------------------------------------------
	// Privates
	// -------------------------------------------------------------------------
	/**
	 * @return
	 */
	private SearchCondition getCondition() {

		Intent intent = getIntent();
		SearchCondition searchCondition = intent.getParcelableExtra(INTENTEXTRA_CONDITION);
		if (searchCondition == null) {
			throw new RuntimeException("Please set extra data by Intent#putExtra(" + INTENTEXTRA_CONDITION + ", Parcelable)");
		}
		return searchCondition;
	}

	private AppPermissionsFinder.SearchCondition toAppPermissionsFinderCondition(SearchCondition cond) {

		AppPermissionsFinder.SearchCondition ret = new AppPermissionsFinder.SearchCondition();
		ret.includeSystemApps = cond.includeSystemApps;
		ret.permissionNamePatternList = cond.permissionNamePatternList;
		ret.protectionLevelSet = new HashSet<Integer>(cond.protectionLevelList);
		return ret;
	}

	/**
	 * @param resources
	 * @param list
	 */
	private void showHitCount(Resources resources, List<PermissionInfoEx> list) {

		TextView textView = (TextView) findViewById(R.id.permission_list_hit_count);
		textView.setText(resources.getString(R.string.hit_count, list.size()));
	}

	/**
	 * @param list
	 */
	private void showList(List<PermissionInfoEx> list) {
		ListView listView = (ListView) findViewById(R.id.permission_list_list);
		adaptor = new PermissionListAdaptor(this, R.layout.permission_list_item, list);
		listView.setAdapter(adaptor);
	}

	// -------------------------------------------------------------------------
	// Classes
	// -------------------------------------------------------------------------
	/**
	 * 
	 * @author takao
	 */
	public static class SearchCondition implements Parcelable {

		public boolean includeSystemApps;
		public List<String> permissionNamePatternList;
		public List<Integer> protectionLevelList;

		public SearchCondition() {
		}

		/**
		 * @param parcel
		 */
		private SearchCondition(Parcel parcel) {
			boolean[] bools = new boolean[1];
			parcel.readBooleanArray(bools);
			includeSystemApps = bools[0];
			permissionNamePatternList = new ArrayList<String>();
			parcel.readList(permissionNamePatternList, null);
			protectionLevelList = new ArrayList<Integer>();
			parcel.readList(protectionLevelList, null);
		}

		/*
		 * (non-Javadoc)
		 * 
		 * @see android.os.Parcelable#describeContents()
		 */
		@Override
		public int describeContents() {
			return 0;
		}

		/*
		 * (non-Javadoc)
		 * 
		 * @see android.os.Parcelable#writeToParcel(android.os.Parcel, int)
		 */
		@Override
		public void writeToParcel(Parcel dest, int flags) {
			dest.writeBooleanArray(new boolean[] { includeSystemApps });
			dest.writeList(permissionNamePatternList);
			dest.writeList(protectionLevelList);
		}

		/**
		 * 
		 */
		public static final Parcelable.Creator<SearchCondition> CREATOR = new Creator<SearchCondition>() {

			/*
			 * (non-Javadoc)
			 * 
			 * @see android.os.Parcelable.Creator#newArray(int)
			 */
			@Override
			public SearchCondition[] newArray(int i) {
				return new SearchCondition[i];
			}

			/*
			 * (non-Javadoc)
			 * 
			 * @see
			 * android.os.Parcelable.Creator#createFromParcel(android.os.Parcel)
			 */
			@Override
			public SearchCondition createFromParcel(Parcel parcel) {
				return new SearchCondition(parcel);
			}
		};
	}
}

File: app/src/main/kotlin/sakaitakao/android/permissionsearch/activity/PermissionListActivity.kt
package sakaitakao.android.permissionsearch.activity


import android.app.Activity
import android.content.Intent
import android.content.res.Resources
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import kotlinx.android.synthetic.main.permission_list.*
import sakaitakao.android.permissionsearch.R
import sakaitakao.android.permissionsearch.adaptor.PermissionListAdaptor
import sakaitakao.android.permissionsearch.entity.PermissionInfoEx
import sakaitakao.android.permissionsearch.model.AppPermissionsFinder
import java.util.*


/**
 * 権限一覧のアクティビティ

 * @author takao
 */
class PermissionListActivity : Activity() {


    /** Called when the activity is first created.  */
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.permission_list)

        val condition = condition

        val packageManager = packageManager
        val aph = AppPermissionsFinder()
        val list = aph.getPermissionInfoExList(packageManager, toAppPermissionsFinderCondition(condition))

        val resources = resources
        showHitCount(resources, list)
        showList(list)
    }

    // -------------------------------------------------------------------------
    // Privates
    // -------------------------------------------------------------------------
    /**
     * @return
     */
    private val condition: SearchCondition
        get() {
            val searchCondition = intent.getParcelableExtra<SearchCondition>(INTENTEXTRA_CONDITION) ?:
                    throw RuntimeException("Please set extra data by Intent#putExtra($INTENTEXTRA_CONDITION, Parcelable)")
            return searchCondition
        }

    private fun toAppPermissionsFinderCondition(cond: SearchCondition): AppPermissionsFinder.SearchCondition {

        val ret = AppPermissionsFinder.SearchCondition()
        ret.includeSystemApps = cond.includeSystemApps
        ret.permissionNamePatternList = cond.permissionNamePatternList.toMutableList()
        ret.protectionLevelSet = HashSet(cond.protectionLevelList)
        return ret
    }

    /**
     * @param resources
     * *
     * @param list
     */
    private fun showHitCount(resources: Resources, list: List<PermissionInfoEx>) {

        permission_list_hit_count.text = resources.getString(R.string.hit_count, list.size)
    }

    /**
     * @param list
     */
    private fun showList(list: List<PermissionInfoEx>) {

        val adaptor = PermissionListAdaptor(this, R.layout.permission_list_item, list)
        permission_list_list.adapter = adaptor
    }

    // -------------------------------------------------------------------------
    // Classes
    // -------------------------------------------------------------------------
    /**

     * @author takao
     */
    class SearchCondition : Parcelable {

        val includeSystemApps: Boolean
        val permissionNamePatternList: List<String>
        val protectionLevelList: List<Int>?

        constructor() {
            includeSystemApps = false
            permissionNamePatternList = ArrayList<String>()
            protectionLevelList = ArrayList<Int>()
        }

        constructor(includeSystemApps: Boolean, permissionNamePatternList: List<String>, protectionLevelList: List<Int>?) {
            this.includeSystemApps = includeSystemApps
            this.permissionNamePatternList = permissionNamePatternList
            this.protectionLevelList = protectionLevelList
        }

        /**
         * @param parcel
         */
        private constructor(parcel: Parcel) {
            val bools = BooleanArray(1)
            parcel.readBooleanArray(bools)
            includeSystemApps = bools[0]
            permissionNamePatternList = ArrayList<String>()
            parcel.readList(permissionNamePatternList, null)
            protectionLevelList = ArrayList<Int>()
            parcel.readList(protectionLevelList, null)
        }

        /*
		 * (non-Javadoc)
		 * 
		 * @see android.os.Parcelable#describeContents()
		 */
        override fun describeContents(): Int {
            return 0
        }

        /*
		 * (non-Javadoc)
		 * 
		 * @see android.os.Parcelable#writeToParcel(android.os.Parcel, int)
		 */
        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeBooleanArray(booleanArrayOf(includeSystemApps))
            dest.writeList(permissionNamePatternList)
            dest.writeList(protectionLevelList)
        }

        companion object {

            @JvmField
            val CREATOR: Parcelable.Creator<SearchCondition> = object : Parcelable.Creator<SearchCondition> {
                override fun newArray(i: Int): Array<SearchCondition?> {
                    return arrayOfNulls(i)
                }

                override fun createFromParcel(parcel: Parcel): SearchCondition {
                    return SearchCondition(parcel)
                }
            }
        }
    }

    companion object {

        private val INTENTEXTRA_CONDITION = PermissionListActivity::class.java.name + ".CONDITION"

        /**
         * アクティビティ起動時の検索条件を設定する。

         * @param intent
         * *            [Intent]
         * *
         * @param searchCondition
         * *            検索条件
         */
        fun setSearchCondition(intent: Intent, searchCondition: SearchCondition) {
            intent.putExtra(PermissionListActivity.INTENTEXTRA_CONDITION, searchCondition)
        }
    }
}